compiler/rustc_middle/src/ty/context.rs

```diff
@@ -2866,7 +2872,24 @@ impl<'tcx> TyCtxt<'tcx> {
             substs.collect::<Vec<_>>(),
         );
         let substs = self.mk_substs(substs);
-        ty::TraitRef::new(trait_def_id, substs)
+        ty::TraitRef { def_id: trait_def_id, substs, _use_mk_trait_ref_instead: () }
+    }
+
+    pub fn mk_projection_ty(
+        self,
+        item_def_id: DefId,
+        substs: impl IntoIterator<Item = impl Into<GenericArg<'tcx>>>,
+    ) -> ty::ProjectionTy<'tcx> {
+        let substs = substs.into_iter().map(Into::into);
+        let n = self.generics_of(item_def_id).count();
+        debug_assert_eq!(
+            (n, Some(n)),
+            substs.size_hint(),
+            "wrong number of generic parameters for {item_def_id:?}: {:?} \nDid you accidentally include the self-type in the params list?",
```

Comment:
same here, the self-type is part of the substs, so that comment feels wrong :thinking: